crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs

```diff
@@ -0,0 +1,800 @@
+                    }
+
+                    let _ = result_sender.send(tx_result); // If the receiver was dropped, ignore
+
+                    let user_nonce = queue
+                        .non_persisted
+                        .user_nonce()
+                        .expect("We just marked the nonce on the queue, so it cannot be None");
+                    println!("\nAfter execution, next nonce being queued is {user_nonce}");
+                    loop {
+                        let Some(head_entry) = queue.txs.first_entry() else {
+                            break;
+                        };
+                        match head_entry.key().cmp(&user_nonce) {
+                            Ordering::Less => {
+                                // Stale transaction in queue - evict and ignore.
+                                // This should not normally happen either, but we handle it to avoid a deadlock
+                                // if it does happen for any reason.
+                                head_entry.remove();
+                                continue;
+                            }
+                            Ordering::Greater => {
+                                // First transaction starts in the future - nothing ready to execute yet
+                                break;
+                            }
+                            Ordering::Equal => {
+                                // First transaction is the next expected nonce. Pop it and send
+                                // as a NewTx.
+                                let tx = head_entry.remove();
+                                queue.non_persisted.mark_inflight(user_nonce);
+                                let _ = self
+                                    .input_sender
+                                    .buffer_sender_channel
+                                    .send(NonceBufferInput::NewTx {
+                                        credential_id,
+                                        baked_tx: tx.baked_tx,
+                                        tx_hash: tx.tx_hash,
+                                        tx_nonce: user_nonce,
+                                        original_tx_queue_id: tx.original_tx_queue_id,
+                                        result_sender: tx.result_sender,
+                                    })
+                                    .await;
+                            }
+                        }
+                    }
+                }
+                NonceBufferInput::TxPersisted { credential_id } => {
+                    let hash_map::Entry::Occupied(entry) = self.buffers.entry(credential_id) else {
+                        continue;
+                    };
+                    if entry.get().txs.is_empty() {
+                        let real_nonce = self
+                            .execution_backend
+                            .get_current_nonce_for_user(&credential_id);
```

Comment:
nit: I think we can avoid the problems of asynchrony between the backend and the tx queue here by just putting the executed nonce into the `TxPersisted` message. We already have access to it when we create the message, and it'll also let us save the state read

Reply:
I had it that way at first but had decided it'd be easier to just check the real nonce once - if txs are inserted in batches into the DB, the TxPersisted messages could come all at once in arbitrary order, and just reading the state lets us clean up the queue once and the rest will be no-ops. Doing it based on the message does save the state read though you're right.

But can you clarify what asynchrony you're worried about here? I don't see any issues other than the state read optimisation, so I wanna make sure I'm not missing anything